app/services/guest-api.ts

```diff
@@ -104,84 +108,117 @@ export class GuestApiService extends Service {
+
+  /**
+   * Extracts mapped args from a request.  Placeholders for callbacks
+   * will be replaced with actual functions.
+   * @param request the request object
+   * @param contents the calling webcontents
+   */
+  private getMappedArgs(request: IGuestApiRequest, contents: electron.WebContents): any[] {
+    return request.args.map(arg => {
+      const isCallbackPlaceholder = typeof arg === 'object' && arg && arg.__guestApiCallback;
+
+      if (isCallbackPlaceholder) {
+        return (...args: any[]) => {
+          const callbackObj: IGuestApiCallback = {
+            args,
+            requestId: request.id,
+            callbackId: arg.id,
+          };
+
+          this.safeSend(contents, 'guestApiCallback', callbackObj);
+        };
+      }
+
+      return arg;
+    });
+  }
+
+  /**
+   * Sends a message stating that the requested endpoint does not exist
+   * @param request the request object
+   * @param contents the calling webcontents
+   */
+  private handleMissingEndpoint(request: IGuestApiRequest, contents: electron.WebContents) {
+    const response: IGuestApiResponse = {
+      id: request.id,
+      error: true,
+      result: `Error: The function ${request.methodPath.join('.')} does not exist!`,
+      resultProcessing: EResponseResultProcessing.None,
+    };
+    this.safeSend(contents, 'guestApiResponse', response);
+  }
+
+  /**
+   * Call an endpoint method, and handle proxying the response
+   * back to the calling webcontents.
+   * @param method The endpoint method to call
+   * @param args The args to call with
+   * @param request The request object
+   * @param contents The calling webcontents
+   */
+  private callEndpointMethod(
+    method: RequestHandlerMethod,
+    args: any[],
+    request: IGuestApiRequest,
+    contents: electron.webContents,
+  ) {
+    method(...args)
+      .then(result => {
+        let response: IGuestApiResponse;
+
+        if (result instanceof FileReturnWrapper) {
+          response = {
+            result: result.filePath,
+            resultProcessing: EResponseResultProcessing.File,
+            id: request.id,
+            error: false,
+          };
+        } else {
+          response = {
+            result,
+            resultProcessing: EResponseResultProcessing.None,
+            id: request.id,
+            error: false,
+          };
+        }
+
+        this.safeSend(contents, 'guestApiResponse', response);
+      })
+      .catch(rawResult => {
+        const result = rawResult instanceof Error ? rawResult.message : rawResult;
+
+        const response: IGuestApiResponse = {
+          result,
+          resultProcessing: EResponseResultProcessing.None,
+          id: request.id,
+          error: true,
+        };
+
+        this.safeSend(contents, 'guestApiResponse', response);
+      });
   }
 
-  private safeSend(contents: Electron.WebContents, channel: string, msg: any) {
+  private safeSend(contents: Electron.WebContents, channel: string, msg?: any) {
```

Comment:
Do we have limited options for `channel`? If so I would have a Enum or restricted string type;

Reply:
Electron doesn't restrict them, but we should probably have all of our IPC messages enumerated somewhere.  Seems like a good refactor for another day.